### PR TITLE
`parse_factor()` now can accept `levels = NULL`

### LINKED ISSUE #497: col_factor() that automatically figures out levels
A nice feature to the `col_factor()` function in `col_types` would be to allow empty arguments. 
This should be treated similar to `data$variable <- factor(data$variable)` on character variables after using `read_csv`. That is, `col_factor()` would be `stringsAsFactors=TRUE` for the specific variables/columns.

### PR DESCRIPTION
**Pull request by jimhester**
This allows one to generate factor levels based on the data

Fixes #497

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # readr 1.0.0.9000
 
+* `parse_time()` now can accept `levels = NULL`, which allows one to generate factor levels based on the data (like stringsAsFactors = TRUE) (#497).
```
Comment (hadley) [inline NEWS.md]: I think you mean `parse_factor()` 😉

File: R/collectors.R [hunk 2]
```diff
@@ -198,7 +198,8 @@ guess_parser <- function(x, locale = default_locale()) {
 #' `parse_factor` is similar to [factor()], but will generate
 #' warnings if elements of `x` are not found in `levels`.
 #'
-#' @param levels Character vector providing set of allowed levels.
+#' @param levels Character vector providing set of allowed levels. if `NULL`,
+#' will generate levels based on the unique values of `x`.
```
Comment (hadley) [inline R/collectors.R]: Need indent

File: R/collectors.R [hunk 3]
```diff
@@ -216,6 +217,9 @@ guess_parser <- function(x, locale = default_locale()) {
 #'
 #' # parse_factor generates a warning & problems
 #' x2 <- parse_factor(x, levels)
+#'
+#' # Using an argument of `NULL` will generate levels based on values of `x`
+#' x2 <- parse_factor(x, levels = NULL)
```
Comment (hadley) [inline R/collectors.R]: How will the levels be ordered?

Comment (jimhester) [inline R/collectors.R]: I guess it is implementation dependant based on std::map, how should we order them, just sorted alphabetically?

Comment (jimhester) [inline R/collectors.R]: According to SO we can depend on it being alphabetically http://stackoverflow.com/questions/7648756/is-the-order-of-iterating-through-stdmap-known-and-guaranteed-by-the-standard

Comment (jimhester) [inline R/collectors.R]: And actually you are right, they are ordered by how they appear in the input, forgot how I implemented this.

**Comment by jimhester** [comment]
Yes that is right, the levels will be ordered how they appear in the input.

**Comment by jimhester** [comment]
Closing in favor of https://github.com/tidyverse/readr/pull/617

**Pull request closed without merging**